File: figures/figure2_forest_dose_response.py
"""
Figure 2 (Python) -- forest plot of effect estimates and engagement exposure-level.

Panel A: primary estimates (ITT, PP), sensitivity (excluding Study Week 3), and
the instrumental-variable CACE (Wald and Anderson-Rubin).
Panel B: effect by engagement level (active engagement, technical failure,
passive exposure).

Reproduces: Supp Fig 2A (forest / exposure-level). All plotted values are fixed
from the reported estimates and asserted below.

Inputs : none (values are fixed constants, asserted)
Outputs: results/figures/Figure2_v14_*.png and Figure2_v14.pdf
"""

import sys
import os

try:
    import matplotlib.pyplot as plt
    from matplotlib.gridspec import GridSpec
    from matplotlib.transforms import blended_transform_factory
    from matplotlib.lines import Line2D
    import numpy as np
except ImportError as e:
    print(f"ERROR: Missing package: {e}")
    sys.exit(1)

# Output directory -- repo root resolved from this file's location
# (figures/ -> repo root) so the script runs unchanged after a fresh clone.
BASE = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
OUTPUT_DIR = os.path.join(BASE, "results", "figures")
os.makedirs(OUTPUT_DIR, exist_ok=True)

# ============================================================================
# STYLE DEFINITIONS
# ============================================================================

plt.rcParams["font.sans-serif"] = ["Arial", "Helvetica", "DejaVu Sans"]
plt.rcParams["font.family"] = "sans-serif"
plt.rcParams["pdf.fonttype"] = 42
plt.rcParams["ps.fonttype"] = 42
plt.rcParams["svg.fonttype"] = "none"

COLORS = {
    "primary":   "#333333",   # Unified color for all estimates and CIs
    "zero":      "#AAAAAA",   # zero reference line
    "header":    "#000000",   # group header text
    "rule":      "#CCCCCC",   # separator lines
    "rule_dark": "#999999",   # group separator lines
}

FONTS = {
    "panel_title":  10.5,
    "group_header": 9.5,
    "subheader":    7.5,
    "row_label":    8,
    "data":         8,
    "col_header":   8,
    "axis":         9,
}

MARKER_SIZE = 5.0

GROUP_GAP = 0.4
HEADER_OFFSET = 0.70
SUBHEADER_EXTRA_GAP = 0.30
SUBHEADER_OFFSET = 0.55

# ============================================================================
# DATA (fixed effect estimates)
# ============================================================================

GROUPS = {
    0: "Primary Estimates",
    1: "Sensitivity",
    2: "Instrumental variable",
    3: "Exposure-level"
}

SUBHEADERS = {}

# Dagger footnote marker (Unicode U+2020) for the per-protocol rows, written
# as a pure-ASCII escape so the source stays ASCII while the figure renders it.
DAGGER = "\u2020"
ROWS = [
    # Group 0: Primary Estimates (Panel A start)
    {"label": "Intention-to-treat",              "est": -9.4,  "lo": -19.9, "hi":  1.1, "p": 0.077,  "n1": 584, "n2": 554, "group": 0, "shape": "o", "fill": "full"},
    {"label": "Per-protocol " + DAGGER,          "est": -13.3, "lo": -26.2, "hi": -0.4, "p": 0.043,  "n1": 259, "n2": 554, "group": 0, "shape": "o", "fill": "full"},
    # Group 1: Sensitivity
    {"label": "ITT excl. Week 3",                "est": -20.1, "lo": -38.1, "hi": -2.0, "p": 0.001,  "n1": 444, "n2": 423, "group": 1, "shape": "o", "fill": "full"},
    {"label": "PP excl. Week 3 " + DAGGER,       "est": -24.2, "lo": -48.0, "hi": -0.5, "p": 0.001,  "n1": 206, "n2": 423, "group": 1, "shape": "o", "fill": "full"},
    # Group 2: Instrumental variable
    {"label": "CACE (Wald LATE)",                "est": -32.2, "lo": -65.2, "hi":  0.9, "p": 0.056,  "n1": 259, "n2": 0,   "group": 2, "shape": "D", "fill": "full"},
    {"label": "CACE (Anderson-Rubin)",           "est": -32.2, "lo": -65.6, "hi":  0.9, "p": 0.057,  "n1": 259, "n2": 0,   "group": 2, "shape": "D", "fill": "full"},
    # Group 3: Exposure-level (Panel B)
    {"label": "Active engagement",               "est": -13.6, "lo": -26.4, "hi": -0.4, "p": 0.048,  "n1": 215, "n2": 554, "group": 3, "shape": "o", "fill": "full"},
    {"label": "Technical failure",               "est": -11.5, "lo": -39.1, "hi": 20.9, "p": 0.45,   "n1":  44, "n2": 554, "group": 3, "shape": "o", "fill": "full"},
    {"label": "Passive exposure",                "est": -1.3,  "lo": -14.2, "hi": 12.3, "p": 0.84,   "n1": 273, "n2": 554, "group": 3, "shape": "o", "fill": "full"},
]

# ============================================================================
# ASSERTIONS (verify fixed values)
# ============================================================================

assert ROWS[0]["est"] == -9.4,   "ITT estimate"
assert ROWS[1]["est"] == -13.3,  "PP estimate"
assert ROWS[2]["est"] == -20.1,  "ITT excl Week 3 estimate"
assert ROWS[3]["est"] == -24.2,  "PP excl Week 3 estimate"
assert ROWS[4]["est"] == -32.2,  "CACE Wald estimate"
assert ROWS[5]["est"] == -32.2,  "CACE AR estimate"
assert ROWS[6]["est"] == -13.6,  "Active estimate"
assert ROWS[6]["p"]   == 0.048,  "Active p-value"
assert ROWS[7]["est"] == -11.5,  "TechFailure estimate"
assert ROWS[8]["est"] == -1.3,   "Passive estimate"
assert ROWS[8]["p"]   == 0.84,   "Passive p-value"

print("[OK] Data assertions passed (9 rows, 4 groups)")

# ============================================================================
# HELPERS
# ============================================================================

def format_p(p):
    """Format a p-value with a leading zero for decimal alignment."""
    if p is None:
        return ""
    if p < 0.001:
        return "<0.001"
    return f"{p:.3f}"


def compute_layout():
    """Compute exact y positions for all rows, panel elements, and headers."""
    layout = {}
    row_ys = []
    group_header_ys = {}

    y = 0.0
    layout['global_rule_top'] = y
    y += 0.5
    layout['global_headers'] = y
    y += 0.5
    layout['global_rule_bottom'] = y
    y += 0.6

    layout['panel_a_title'] = y
    y += 0.8

    current_group = -1
    for i, row in enumerate(ROWS):
        g = row["group"]

        if g == 3 and current_group != 3:
            # Mirror the Panel A title spacing for the Panel B title.
            y += 0.3
            layout['panel_b_title'] = y
            y += 0.8

        if g != current_group:
            if current_group >= 0 and g != 3:  # Intra-panel gaps
                y += GROUP_GAP
            group_header_ys[g] = y
            y += HEADER_OFFSET
            current_group = g

        row_ys.append(y)
        y += 1.0

    y -= 0.6  # Adjust bottom padding
    layout['panel_b_rule_end'] = y

    return layout, row_ys, group_header_ys


# ============================================================================
# MAIN FIGURE
# ============================================================================

COL_N = 0.12        # N: center-aligned
COL_EST = 0.37      # Effect: right-aligned
COL_CI = 0.62       # 95% CI: center-aligned
COL_P = 0.92        # P Value: right-aligned

def draw_column_headers(ax_label, ax_data, y_pos):
    """Draw column headers at the given Y position for the Label and Data columns."""
    ax_label.text(0.02, y_pos, "Analysis",
                  fontsize=FONTS["col_header"], fontweight="bold",
                  va="center", ha="left",
                  transform=ax_label.get_yaxis_transform())

    ax_data.text(COL_N, y_pos, "N\n(Interv./Control)",
                 fontsize=FONTS["col_header"], fontweight="bold",
                 va="center", ha="center", linespacing=0.95,
                 transform=ax_data.get_yaxis_transform())
    ax_data.text(COL_EST, y_pos, "Effect\n(min)",
                 fontsize=FONTS["col_header"], fontweight="bold",
                 va="center", ha="center", linespacing=0.95,
                 transform=ax_data.get_yaxis_transform())
    ax_data.text(COL_CI, y_pos, "95% CI",
                 fontsize=FONTS["col_header"], fontweight="bold",
                 va="center", ha="center", linespacing=0.95,
                 transform=ax_data.get_yaxis_transform())
    ax_data.text(COL_P, y_pos, "P Value",
                 fontsize=FONTS["col_header"], fontweight="bold",
                 va="center", ha="right", linespacing=0.95,
                 transform=ax_data.get_yaxis_transform())

def draw_header_rules(axes, y_top, y_bottom):
    """Draw sandwich header rules across all axes (thick top, thin bottom)."""
    for ax in axes:
        ax.axhline(y=y_top, color="black", linewidth=1.5,
                   clip_on=False, solid_capstyle="butt")
        ax.axhline(y=y_bottom, color="black", linewidth=0.8,
                   clip_on=False, solid_capstyle="butt")

def create_figure():
    """Create the grouped forest plot."""
    layout, row_ys, group_header_ys = compute_layout()

    fig = plt.figure(figsize=(7.1, 7.8))

    gs = GridSpec(1, 3, figure=fig, width_ratios=[0.33, 0.27, 0.40],
                  wspace=0.0)
    ax_label = fig.add_subplot(gs[0, 0])
    ax_forest = fig.add_subplot(gs[0, 1], sharey=ax_label)
    ax_data = fig.add_subplot(gs[0, 2], sharey=ax_label)

    y_min = layout['global_rule_top'] - 0.5
    y_max = layout['panel_b_rule_end'] + 0.6

    for ax in (ax_label, ax_forest, ax_data):
        ax.set_ylim(y_min, y_max)
        ax.invert_yaxis()

    all_axes = [ax_label, ax_forest, ax_data]

    # ------------------------------------------------------------------
    # RULES & SEPARATORS
    # ------------------------------------------------------------------
    draw_header_rules(all_axes, layout['global_rule_top'], layout['global_rule_bottom'])

    # ------------------------------------------------------------------
    # COLUMN 1: Labels & Titles
    # ------------------------------------------------------------------
    ax_label.set_xlim(0, 1)
    ax_label.axis("off")

    ax_label.text(0.00, layout['panel_a_title'], "Panel A: Effect Estimates",
                  fontsize=FONTS["panel_title"], fontweight="bold",
                  va="center", ha="left", transform=ax_label.get_yaxis_transform())

    ax_label.text(0.00, layout['panel_b_title'], "Panel B: Effect by Engagement",
                  fontsize=FONTS["panel_title"], fontweight="bold",
                  va="center", ha="left", transform=ax_label.get_yaxis_transform())

    for g, hy in group_header_ys.items():
        ax_label.text(0.03, hy, GROUPS[g],
                      fontsize=FONTS["group_header"], fontweight="bold",
                      va="center", ha="left", transform=ax_label.get_yaxis_transform())

    for i, ry in enumerate(row_ys):
        indent = 0.06
        ax_label.text(indent, ry, ROWS[i]["label"],
                      fontsize=FONTS["row_label"], va="center", ha="left",
                      transform=ax_label.get_yaxis_transform())

    # ------------------------------------------------------------------
    # COLUMN 2: Forest plot
    # ------------------------------------------------------------------
    x_lo, x_hi = -70, 25
    ax_forest.set_xlim(x_lo, x_hi)
    ax_forest.set_xticks([-60, -40, -20, 0, 20])
    ax_forest.spines["top"].set_visible(False)
    ax_forest.spines["right"].set_visible(False)
    ax_forest.spines["left"].set_visible(False)
    ax_forest.set_yticks([])
    ax_forest.spines["bottom"].set_color("black")
    ax_forest.spines["bottom"].set_linewidth(1.0)
    ax_forest.tick_params(axis="x", labelsize=FONTS["data"], width=1.0,
                          length=4, pad=3)

    ax_forest.plot([0, 0], [layout['global_rule_bottom'], layout['panel_b_rule_end']],
                   color=COLORS["zero"], linewidth=0.8, zorder=1)

    ax_forest.set_xlabel("Consultation time difference (minutes)",
                         fontsize=FONTS["axis"], fontweight="bold",
                         labelpad=6)

    for i, ry in enumerate(row_ys):
        row = ROWS[i]
        est, lo, hi = row["est"], row["lo"], row["hi"]
        color = COLORS["primary"]

        lo_draw = max(lo, x_lo)
        hi_draw = min(hi, x_hi)

        ax_forest.plot([lo_draw, hi_draw], [ry, ry],
                       color=color, linewidth=1.8,
                       linestyle="-", zorder=5,
                       solid_capstyle="round")

        cap_half = 0.15
        if lo >= x_lo:
            ax_forest.plot([lo, lo], [ry - cap_half, ry + cap_half],
                           color=color, linewidth=1.5, zorder=5,
                           solid_capstyle="round")
        if hi <= x_hi:
            ax_forest.plot([hi, hi], [ry - cap_half, ry + cap_half],
                           color=color, linewidth=1.5, zorder=5,
                           solid_capstyle="round")

        marker_edge = color
        marker_face = color if row["fill"] == "full" else "white"

        ax_forest.plot(est, ry, row["shape"],
                       color=color, markerfacecolor=marker_face,
                       markeredgecolor=marker_edge, markeredgewidth=1.2,
                       markersize=MARKER_SIZE, zorder=6)

    # ------------------------------------------------------------------
    # COLUMN 3: Data columns / Headers
    # ------------------------------------------------------------------
    ax_data.set_xlim(0, 1)
    ax_data.axis("off")

    draw_column_headers(ax_label, ax_data, layout['global_headers'])

    for i, ry in enumerate(row_ys):
        row = ROWS[i]

        n1, n2 = row['n1'], row['n2']
        n_str = f"{n1}/{n2}" if n2 > 0 else f"{n1}"
        ax_data.text(COL_N, ry, n_str,
                     fontsize=FONTS["data"], va="center", ha="center", transform=ax_data.get_yaxis_transform())

        est_str = f"{row['est']:.1f}"
        ax_data.text(COL_EST + 0.03, ry, est_str,
                     fontsize=FONTS["data"], va="center", ha="right", transform=ax_data.get_yaxis_transform())

        ci_str = f"{row['lo']:.1f} to {row['hi']:.1f}"
        ax_data.text(COL_CI, ry, ci_str,
                     fontsize=FONTS["data"], va="center", ha="center", transform=ax_data.get_yaxis_transform())

        p_str = format_p(row["p"])
        ax_data.text(COL_P, ry, p_str,
                     fontsize=FONTS["data"], va="center", ha="right",
                     color="#000000", fontweight="normal", transform=ax_data.get_yaxis_transform())

    legend_elements = [
        Line2D([0], [0], marker='o', color='w', markerfacecolor=COLORS["primary"],
               markersize=MARKER_SIZE, label='Median based estimate (HL)'),
        Line2D([0], [0], marker='D', color='w', markerfacecolor=COLORS["primary"],
               markersize=MARKER_SIZE, label='Mean based estimate')
    ]
    ax_data.legend(handles=legend_elements, loc='upper right', ncol=1,
                   frameon=False, edgecolor='none', framealpha=1,
                   fontsize=FONTS["row_label"], borderaxespad=0.0,
                   bbox_to_anchor=(COL_P, -0.05))

    plt.subplots_adjust(left=0.02, right=0.98, top=0.94, bottom=0.18)

    return fig


# ============================================================================
# MAIN
# ============================================================================

def main():
    mode = sys.argv[1] if len(sys.argv) > 1 else "preview"
    print(f"[INFO] Figure 2 run mode: {mode}")

    fig = create_figure()

    base = os.path.join(OUTPUT_DIR, "Figure2_v14")

    if mode == "preview":
        path = f"{base}_preview.png"
        fig.savefig(path, dpi=180, bbox_inches="tight",
                    facecolor="white", edgecolor="none")
        print(f"Saved: {path}")
    elif mode == "final":
        for dpi in [300, 600]:
            path = f"{base}_{dpi}dpi.png"
            fig.savefig(path, dpi=dpi, bbox_inches="tight",
                        facecolor="white", edgecolor="none")
            print(f"Saved: {path}")
        pdf_path = f"{base}.pdf"
        fig.savefig(pdf_path, bbox_inches="tight",
                    facecolor="white", edgecolor="none")
        print(f"Saved: {pdf_path}")
    else:
        print(f"Unknown mode: {mode}. Use 'preview' or 'final'.")
        sys.exit(1)

    plt.close(fig)
    print(f"[DONE] Figure 2 complete (mode: {mode})")


if __name__ == "__main__":
    main()

File: figures/figure2_forest_dose_response.R
# =============================================================================
# figure2_forest_dose_response.R -- SHAKED manuscript Figure 2 (R version)
#
# Panel A: Causal estimation hierarchy (ITT / PP / CACE)
#   - Uniform neutral styling; no significance encoding
#   - 3 rows, all in minutes (HL / Wald estimates)
#
# Panel B: Engagement-level mechanism (exposure-level)
#   - Significance-encoded: null -> null -> significant staircase
#   - 4 rows (Wing B reference + 3 engagement levels), HL shifts vs Wing B
#
# Reproduces: Supp Fig 2A (forest / exposure-level). Values are fixed and asserted.
# Output: results/figures/Figure2_Forest_DoseResponse_v3_600dpi.png + .pdf
# Size: 180mm x ~84mm
# =============================================================================

# Resolve _common.R next to this script (portable across CWDs).
.this_file <- sub("^--file=", "", grep("^--file=", commandArgs(FALSE), value = TRUE)[1])
source(file.path(dirname(.this_file), "_common.R"))

# Run modes: preview | final | both (default)
args <- commandArgs(trailingOnly = TRUE)
run_mode <- if (length(args) >= 1) tolower(args[1]) else "both"
if (!run_mode %in% c("preview", "final", "both")) {
    stop("Invalid mode. Use: preview, final, or both")
}
message(sprintf("[INFO] Figure 2 run mode: %s", run_mode))

# =============================================================================
# 1. DATA -- fixed effect estimates
# =============================================================================

# Panel A: 3 causal estimation approaches (all minutes)
panel_a_data <- data.frame(
    label    = c("Intention-to-treat", "Per-protocol", "CACE (IV)"),
    estimate = c(-9.4, -13.3, -32.2),
    ci_low   = c(-19.9, -26.2, -65.2),
    ci_high  = c(1.1, -0.4, 0.9),
    p_value  = c(0.077, 0.043, 0.056),
    n_tx     = c(584, 259, 584),
    n_ctrl   = c(554, 554, 554),
    y_pos    = c(2.4, 1.7, 1),
    is_ref   = c(FALSE, FALSE, FALSE),
    stringsAsFactors = FALSE
)

# Panel B: 4 engagement levels (HL shifts vs Wing B reference)
panel_b_data <- data.frame(
    label    = c("Wing B controls (reference)", "Passive exposure",
                 "Technical failure", "Active engagement"),
    estimate = c(0, -1.3, -11.5, -13.6),
    ci_low   = c(NA, -14.2, -39.1, -26.4),
    ci_high  = c(NA, 12.3, 20.9, -0.4),
    p_value  = c(NA, 0.844, 0.449, 0.048),
    n_tx     = c(NA, 273, 44, 215),
    n_ctrl   = c(554, 554, 554, 554),
    y_pos    = c(3.1, 2.4, 1.7, 1),
    is_ref   = c(TRUE, FALSE, FALSE, FALSE),
    stringsAsFactors = FALSE
)

# =============================================================================
# 2. ASSERTIONS
# =============================================================================
stopifnot(nrow(panel_a_data) == 3)
stopifnot(nrow(panel_b_data) == 4)
stopifnot(panel_a_data$estimate[1] == -9.4)    # ITT
stopifnot(panel_a_data$estimate[2] == -13.3)   # PP
stopifnot(panel_a_data$estimate[3] == -32.2)   # CACE
stopifnot(panel_a_data$p_value[2]  == 0.043)   # PP
stopifnot(panel_a_data$p_value[3]  == 0.056)   # CACE
stopifnot(panel_b_data$estimate[4] == -13.6)   # Active
stopifnot(panel_b_data$p_value[4]  == 0.048)   # Active significant
stopifnot(panel_b_data$is_ref[1]   == TRUE)    # Wing B is reference
message("[OK] Figure 2 data assertions passed")

# =============================================================================
# 3. DERIVED COLUMNS
# =============================================================================

FS <- FONT_SIZES
sig_color     <- pal$active   # "#4472C4"
nsig_color    <- "#555555"    # dark gray markers for non-significant
nsig_ci_color <- "#AAAAAA"   # lighter gray CI lines for non-significant
ref_color     <- pal$control  # "#A5A5A5"
neutral_color <- "#333333"    # Panel A uniform color
header_rule_y <- 3.55         # y-position for header rule in Panel A
row_sp        <- 1            # row spacing unit

# --- Panel A derived ---
panel_a_data$pt_color  <- neutral_color
panel_a_data$n_display <- paste0(panel_a_data$n_tx, "/", panel_a_data$n_ctrl)
panel_a_data$est_text  <- sprintf("%.1f", panel_a_data$estimate)
panel_a_data$ci_text   <- sprintf("%.1f to %.1f",
    panel_a_data$ci_low, panel_a_data$ci_high)
panel_a_data$p_text    <- format_p(panel_a_data$p_value)
panel_a_data$p_color   <- neutral_color  # uniform, no significance encoding

# --- Panel B derived ---
panel_b_data$significant <- !panel_b_data$is_ref &
    !is.na(panel_b_data$p_value) & panel_b_data$p_value < 0.05

panel_b_data$pt_color <- ifelse(panel_b_data$is_ref, ref_color,
    ifelse(panel_b_data$significant, sig_color, nsig_color))

panel_b_data$n_display <- ifelse(panel_b_data$is_ref,
    as.character(panel_b_data$n_ctrl),
    paste0(panel_b_data$n_tx, "/", panel_b_data$n_ctrl))

panel_b_data$est_text <- ifelse(panel_b_data$is_ref, "",
    sprintf("%.1f", panel_b_data$estimate))
panel_b_data$ci_text  <- ifelse(panel_b_data$is_ref, "",
    sprintf("%.1f to %.1f", panel_b_data$ci_low, panel_b_data$ci_high))
panel_b_data$p_text   <- ifelse(panel_b_data$is_ref, "",
    format_p(panel_b_data$p_value))
panel_b_data$p_color  <- ifelse(panel_b_data$significant, sig_color, nsig_color)
panel_b_data$p_color[panel_b_data$is_ref] <- nsig_color

panel_b_ci <- panel_b_data[!panel_b_data$is_ref, ]

# Shared x-axis range
x_limits <- c(-75, 25)
x_breaks <- seq(-60, 20, 20)

# =============================================================================
# Helper: build one panel (6-column table-style forest plot)
# =============================================================================
build_panel <- function(df, df_ci, panel_tag, y_range, header_y, rule_y,
                        show_x_axis = FALSE, uniform_style = FALSE) {

    # --- Column 1: Labels ---
    col_labels <- ggplot(df, aes(x = 0, y = y_pos)) +
        geom_text(aes(label = label), hjust = 0, size = FS$data,
            family = "Arial", color = pal$text) +
        annotate("text", x = -0.02, y = header_y + 0.15, label = panel_tag,
            hjust = 0, fontface = "bold", size = FS$tag, family = "Arial",
            color = pal$text) +
        annotate("segment", x = -0.02, xend = 1.1, y = rule_y, yend = rule_y,
            linewidth = 0.3, color = "gray70") +
        scale_y_continuous(limits = y_range, expand = c(0, 0)) +
        scale_x_continuous(limits = c(-0.02, 1)) +
        theme_void() +
        theme(plot.margin = margin(2, 0, 2, 4))

    # --- Column 2: N (Interv./Control) ---
    col_n <- ggplot(df, aes(x = 0.5, y = y_pos)) +
        geom_text(aes(label = n_display), hjust = 0.5, size = FS$stats,
            family = "Arial", color = pal$text) +
        annotate("text", x = 0.5, y = header_y,
            label = "N\n(Tx/Ctrl)",
            hjust = 0.5, fontface = "bold", size = FS$annotation,
            family = "Arial", color = pal$text, lineheight = 0.9) +
        annotate("segment", x = -0.3, xend = 1.3, y = rule_y, yend = rule_y,
            linewidth = 0.3, color = "gray70") +
        scale_y_continuous(limits = y_range, expand = c(0, 0)) +
        scale_x_continuous(limits = c(-0.3, 1.3)) +
        theme_void() +
        theme(plot.margin = margin(2, 0, 2, 0))

    # --- Column 3: CI forest plot ---
    p_ci <- ggplot() +
        geom_vline(xintercept = 0, color = "gray70", linewidth = 0.5)

    if (any(df$is_ref %in% TRUE)) {
        p_ci <- p_ci +
            geom_point(data = df[df$is_ref, ],
                aes(x = estimate, y = y_pos),
                shape = 18, size = 3.5, color = ref_color)
    }

    if (!is.null(df_ci) && nrow(df_ci) > 0) {
        if (uniform_style) {
            # Panel A: all same color, solid lines, filled squares
            p_ci <- p_ci +
                geom_segment(data = df_ci,
                    aes(x = ci_low, xend = ci_high, y = y_pos, yend = y_pos),
                    linewidth = 1.1, color = neutral_color) +
                geom_point(data = df_ci,
                    aes(x = estimate, y = y_pos),
                    shape = 15, size = 3.2, color = neutral_color)
        } else {
            # Panel B: significance-encoded
            nsig_rows <- df_ci[!df_ci$significant, ]
            sig_rows  <- df_ci[df_ci$significant, ]

            if (nrow(nsig_rows) > 0) {
                p_ci <- p_ci +
                    geom_segment(data = nsig_rows,
                        aes(x = ci_low, xend = ci_high,
                            y = y_pos, yend = y_pos),
                        linewidth = 1.0, color = nsig_ci_color) +
                    geom_point(data = nsig_rows,
                        aes(x = estimate, y = y_pos),
                        shape = 15, size = 3.2, color = nsig_color)
            }
            if (nrow(sig_rows) > 0) {
                p_ci <- p_ci +
                    geom_segment(data = sig_rows,
                        aes(x = ci_low, xend = ci_high,
                            y = y_pos, yend = y_pos),
                        linewidth = 1.2, color = sig_color) +
                    geom_point(data = sig_rows,
                        aes(x = estimate, y = y_pos),
                        shape = 15, size = 3.5, color = sig_color)
            }
        }
    }

    p_ci <- p_ci +
        annotate("segment",
            x = x_limits[1], xend = x_limits[2],
            y = rule_y, yend = rule_y,
            linewidth = 0.3, color = "gray70")

    # Direction arrows (only on the bottom panel)
    if (show_x_axis) {
        p_ci <- p_ci +
            annotate("text", x = -38, y = y_range[1] + 0.35,
                label = "\u2190 Favors SHAKED",
                hjust = 0.5, size = FS$annotation, color = "gray50",
                fontface = "italic", family = "Arial") +
            annotate("text", x = 12, y = y_range[1] + 0.35,
                label = "Favors Control \u2192",
                hjust = 0.5, size = FS$annotation, color = "gray50",
                fontface = "italic", family = "Arial")
    }

    p_ci <- p_ci +
        scale_x_continuous(limits = x_limits, breaks = x_breaks) +
        scale_y_continuous(limits = y_range, expand = c(0, 0)) +
        labs(x = if (show_x_axis) "Consultation Time Difference (minutes)"
             else NULL, y = NULL) +
        theme_shaked(base_size = 8) +
        theme(
            axis.text.y  = element_blank(),
            axis.ticks.y = element_blank(),
            axis.text.x  = if (show_x_axis)
                element_text(size = FS$axis_tick * 2.83)
                else element_blank(),
            axis.ticks.x = if (show_x_axis)
                element_line(linewidth = 0.3)
                else element_blank(),
            axis.title.x = if (show_x_axis)
                element_text(size = FS$axis * 2.83)
                else element_blank(),
            panel.grid.major.x = element_line(color = "gray94", linewidth = 0.2),
            panel.grid.major.y = element_blank(),
            plot.margin = margin(2, 2, if (show_x_axis) 5 else 2, 2)
        )

    # --- Column 4: Median diff. ---
    col_est <- ggplot(df, aes(x = 0.5, y = y_pos)) +
        geom_text(aes(label = est_text), hjust = 0.5, size = FS$stats,
            family = "Arial", color = pal$text) +
        annotate("text", x = 0.5, y = header_y,
            label = "Median\ndiff.",
            hjust = 0.5, fontface = "bold", size = FS$annotation,
            family = "Arial", color = pal$text, lineheight = 0.9) +
        annotate("segment", x = -0.3, xend = 1.3, y = rule_y, yend = rule_y,
            linewidth = 0.3, color = "gray70") +
        scale_y_continuous(limits = y_range, expand = c(0, 0)) +
        scale_x_continuous(limits = c(-0.3, 1.3)) +
        theme_void() +
        theme(plot.margin = margin(2, 0, 2, 0))

    # --- Column 5: 95% CI ---
    col_ci_text <- ggplot(df, aes(x = 0.5, y = y_pos)) +
        geom_text(aes(label = ci_text), hjust = 0.5, size = FS$stats,
            family = "Arial", color = pal$text) +
        annotate("text", x = 0.5, y = header_y,
            label = "95% CI",
            hjust = 0.5, fontface = "bold", size = FS$annotation,
            family = "Arial", color = pal$text) +
        annotate("segment", x = -0.3, xend = 1.3, y = rule_y, yend = rule_y,
            linewidth = 0.3, color = "gray70") +
        scale_y_continuous(limits = y_range, expand = c(0, 0)) +
        scale_x_continuous(limits = c(-0.3, 1.3)) +
        theme_void() +
        theme(plot.margin = margin(2, 0, 2, 0))

    # --- Column 6: P Value ---
    col_p <- ggplot(df, aes(x = 0.5, y = y_pos)) +
        geom_text(aes(label = p_text), hjust = 1, size = FS$stats,
            family = "Arial", color = df$p_color) +
        annotate("text", x = 0.5, y = header_y,
            label = "P Value",
            hjust = 1, fontface = "bold", size = FS$annotation,
            family = "Arial", color = pal$text) +
        annotate("segment", x = -0.5, xend = 0.6, y = rule_y, yend = rule_y,
            linewidth = 0.3, color = "gray70") +
        scale_y_continuous(limits = y_range, expand = c(0, 0)) +
        scale_x_continuous(limits = c(-0.5, 0.6)) +
        theme_void() +
        theme(plot.margin = margin(2, 6, 2, 0))

    # Assemble: 6 columns
    (col_labels | col_n | p_ci | col_est | col_ci_text | col_p) +
        plot_layout(widths = c(3.2, 1.2, 2.8, 0.8, 1.6, 0.8))
}

# =============================================================================
# 4. PANEL A -- Uniform styling (no significance encoding)
# =============================================================================

y_range_a  <- c(0.6, 3.15)
header_y_a <- 3.0
rule_y_a   <- 2.88

panel_a <- build_panel(
    df            = panel_a_data,
    df_ci         = panel_a_data,
    panel_tag     = "A",
    y_range       = y_range_a,
    header_y      = header_y_a,
    rule_y        = rule_y_a,
    show_x_axis   = FALSE,
    uniform_style = TRUE
)

# =============================================================================
# 5. PANEL B -- Significance-encoded engagement staircase
# =============================================================================

y_range_b  <- c(0.4, 3.85)
header_y_b <- 3.7
rule_y_b   <- 3.58

panel_b <- build_panel(
    df            = panel_b_data,
    df_ci         = panel_b_ci,
    panel_tag     = "B",
    y_range       = y_range_b,
    header_y      = header_y_b,
    rule_y        = rule_y_b,
    show_x_axis   = TRUE,
    uniform_style = FALSE
)

# =============================================================================
# 6. ASSEMBLY -- stacked vertically, ~40/60 height split
# =============================================================================

fig2 <- (panel_a / panel_b) +
    plot_layout(heights = c(0.40, 0.60))

# =============================================================================
# 7. SAVE
# =============================================================================
FIG_W <- 7.1   # 180mm
FIG_H <- 3.3   # ~84mm

save_preview <- function(plot, name, width, height, dpi = 180) {
    png_path <- file.path(FIG_DIR, paste0(name, "_preview.png"))
    ggsave(
        png_path, plot,
        width = width, height = height, dpi = dpi, device = ragg::agg_png
    )
    cat("Saved:", png_path, "\n")
}

if (run_mode %in% c("preview", "both")) {
    save_preview(fig2, "Figure2_Forest_DoseResponse_v3", width = FIG_W, height = FIG_H)
}
if (run_mode %in% c("final", "both")) {
    save_figure(fig2, "Figure2_Forest_DoseResponse_v3", width = FIG_W, height = FIG_H)
}
cat(sprintf("[DONE] Figure 2 complete (mode: %s).\n", run_mode))
